apps/src/lib/node/ledger/protocol/mod.rs

```diff
@@ -61,56 +59,60 @@ pub enum Error {
     AccessForbidden(InternalAddress),
 }
 
-pub(crate) struct ShellParams<'a, D, H, CA>
-where
-    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
-    H: 'static + StorageHasher + Sync,
-    CA: 'static + WasmCacheAccess + Sync,
-{
-    pub block_gas_meter: &'a mut BlockGasMeter,
-    pub write_log: &'a mut WriteLog,
-    pub storage: &'a Storage<D, H>,
-    pub wasm_dir: &'a std::path::Path,
-    pub vp_wasm_cache: &'a mut VpCache<CA>,
-    pub tx_wasm_cache: &'a mut TxCache<CA>,
-}
+pub type Result<T> = std::result::Result<T, Error>;
 
-impl<'a, D, H> From<&'a mut Shell<D, H>>
-    for ShellParams<'a, D, H, namada::vm::WasmCacheRwAccess>
+/// Dispatch a given transaction to be applied based on its type. Some storage
+/// updates may be derived and applied natively rather than via the wasm
+/// environment, in which case validity predicates will be bypassed.
+///
+/// If the given tx is a successfully decrypted payload apply the necessary
+/// vps. Otherwise, we include the tx on chain with the gas charge added
+/// but no further validations.
+pub(crate) fn dispatch_tx<'a, D, H, CA>(
+    tx_type: TxType,
+    tx_length: usize,
+    block_gas_meter: &'a mut BlockGasMeter,
+    write_log: &'a mut WriteLog,
+    storage: &'a mut Storage<D, H>,
+    vp_wasm_cache: &'a mut VpCache<CA>,
+    tx_wasm_cache: &'a mut TxCache<CA>,
+) -> Result<TxResult>
 where
     D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
     H: 'static + StorageHasher + Sync,
+    CA: 'static + WasmCacheAccess + Sync,
 {
-    fn from(shell: &'a mut Shell<D, H>) -> Self {
-        Self {
-            block_gas_meter: &mut shell.gas_meter,
-            write_log: &mut shell.write_log,
-            storage: &mut shell.storage,
-            wasm_dir: shell.wasm_dir.as_path(),
-            vp_wasm_cache: &mut shell.vp_wasm_cache,
-            tx_wasm_cache: &mut shell.tx_wasm_cache,
+    match tx_type {
+        TxType::Raw(_) => Err(Error::TxTypeError),
+        TxType::Decrypted(DecryptedTx::Decrypted(tx)) => apply_wasm_tx(
+            tx,
+            tx_length,
+            block_gas_meter,
+            write_log,
+            storage,
+            vp_wasm_cache,
+            tx_wasm_cache,
+        ),
+        TxType::Protocol(ProtocolTx { tx, .. }) => {
+            apply_protocol_tx(tx, storage)
+        }
+        _ => {
+            // other transaction types we treat as a noop
+            Ok(TxResult::default())
```

Comment:
I think here we should return `Err(Error::TxTypeError)` rather than noop. The only transaction types we should ever be applying for Namada are decrypted wasm transactions and native protocol transactions.

Reply:
Sure

Reply:
I've opened an issue to look at doing this more in depth as it will break some tests - https://github.com/anoma/namada/issues/313